w_emitter.h

```diff
@@ -538,23 +538,21 @@ class WEmitter
                 std::string size = psize(p, "_args.");
                 std::string cmd = p->attrs_->inout_ ? "OE_READ_IN_OUT_PARAM"
                                                     : "OE_READ_OUT_PARAM";
-                if (p->attrs_->string_ || p->attrs_->wstring_)
-                {
-                    out() << "    " + check +
-                                 (p->attrs_->wstring_ ? "_WIDE(" : "(") +
-                                 "_output_buffer + _output_buffer_offset, "
-                                 "_args." +
-                                 p->name_ + "_len);";
-                }
                 UserType* ut = get_user_type_for_deep_copy(edl_, p);
                 if (!ut)
                 {
                     out() << "    " + cmd + "(" + p->name_ + ", (size_t)(" +
                                  size + "));";
-                    continue;
+                }
```

Comment:
This fix looks good. Can you double check if deep-copy is also affected by this TOCTOU vulnerability?

Reply:
Quick summary: deep-copy is not vulnerable to the similar TOCTOU issue, unless we support `string` attribute.

Please check the example below. We perform the check against in-enclave pointers that the host cannot manipulate. However, I do see the problem that the deep-copy on OCALL allows the host to overwrite the `size` of the edl-annotated struct. I think this requires a patch.

Example:

edl definition:
```
  struct DeepCopy
  {
    size_t size;
    [size=size] char *buf;
  };

  struct NestedDeepCopy
  {
    [count=3] DeepCopy *dp;
    int value;
    size_t size;
    [size=size] char *buf;
  };

...
  void ocall_test([out, count = 2] NestedDeepCopy *ndc, [out, count = 1] DeepCopy *dc);
```

_t.c:
```
    if (ndc) // MS: Check against in-enclave pointer.
    {
        NestedDeepCopy* _rhs = (NestedDeepCopy*) (_output_buffer + _output_buffer_offset); (void)_rhs;
        OE_ADD_SIZE(_output_buffer_offset, (size_t)((size_t)2 * sizeof(NestedDeepCopy)));
        for (size_t _i_1 = 0; _i_1 < 2; _i_1++)
        {
            ndc[_i_1].value = _rhs[_i_1].value; // MS: Controlled by the host.
            ndc[_i_1].size = _rhs[_i_1].size; // MS: Controlled by the host.
            if (ndc[_i_1].dp) // MS: Check against in-enclave pointer.
            {
                DeepCopy* _rhs = (DeepCopy*) (_output_buffer + _output_buffer_offset); (void)_rhs;
                OE_ADD_SIZE(_output_buffer_offset, (size_t)((size_t)3 * sizeof(DeepCopy)));
                for (size_t _i_2 = 0; _i_2 < 3; _i_2++)
                {
                    ndc[_i_1].dp[_i_2].size = _rhs[_i_2].size; // MS: Controlled by the host.
                    OE_READ_OUT_PARAM(ndc[_i_1].dp[_i_2].buf, ndc[_i_1].dp[_i_2].size);
                }
            }
            OE_READ_OUT_PARAM(ndc[_i_1].buf, ndc[_i_1].size);
        }
    }

    if (dc) // MS: Check against in-enclave pointer.
    {
        DeepCopy* _rhs = (DeepCopy*) (_output_buffer + _output_buffer_offset); (void)_rhs;
        OE_ADD_SIZE(_output_buffer_offset, (size_t)((size_t)1 * sizeof(DeepCopy)));
        for (size_t _i_1 = 0; _i_1 < 1; _i_1++)
        {
            dc[_i_1].size = _rhs[_i_1].size; // Controlled by the host.
            OE_READ_OUT_PARAM(dc[_i_1].buf, dc[_i_1].size);
        }
    }
```

Reply:
I think this is a missing front end check.
Deep-copy is not supported for out parameters.
Intel edger8r says:
> error: the structure declaration "S" specifies a deep copy, should not be used with an `out' attribute in function "host_fcn2"

for my example.

Reply:
We need to make sure that the same problem does not occur with in/out parameters.

Reply:
The generated code is the same for the in/out case. However, I'm not sure what's the right fix for this problem. In particular, the attribute is applied to the entire struct. In this case, overwriting any of the struct member seems to be legit. Should we prevent the struct member that is used by `size` or `count` attribute from overwriting?  

```
    /* Unmarshal return value and out, in-out parameters. */
    /* No return value. */
    OE_READ_OUT_PARAM(str1, (size_t)(_args.str1_len * sizeof(char)));
    OE_CHECK_NULL_TERMINATOR(str1, _args.str1_len);
    if (ndc)
    {
        NestedDeepCopy* _rhs = (NestedDeepCopy*) (_output_buffer + _output_buffer_offset); (void)_rhs;
        OE_ADD_SIZE(_output_buffer_offset, (size_t)((size_t)2 * sizeof(NestedDeepCopy)));
        for (size_t _i_1 = 0; _i_1 < 2; _i_1++)
        {
            ndc[_i_1].value = _rhs[_i_1].value;
            ndc[_i_1].size = _rhs[_i_1].size;
            if (ndc[_i_1].dp)
            {
                DeepCopy* _rhs = (DeepCopy*) (_output_buffer + _output_buffer_offset); (void)_rhs;
                OE_ADD_SIZE(_output_buffer_offset, (size_t)((size_t)3 * sizeof(DeepCopy)));
                for (size_t _i_2 = 0; _i_2 < 3; _i_2++)
                {
                    ndc[_i_1].dp[_i_2].size = _rhs[_i_2].size;
                    OE_READ_IN_OUT_PARAM(ndc[_i_1].dp[_i_2].buf, ndc[_i_1].dp[_i_2].size);
                }
            }
            OE_READ_IN_OUT_PARAM(ndc[_i_1].buf, ndc[_i_1].size);
        }
    }
    if (dc)
    {
        DeepCopy* _rhs = (DeepCopy*) (_output_buffer + _output_buffer_offset); (void)_rhs;
        OE_ADD_SIZE(_output_buffer_offset, (size_t)((size_t)1 * sizeof(DeepCopy)));
        for (size_t _i_1 = 0; _i_1 < 1; _i_1++)
        {
            dc[_i_1].size = _rhs[_i_1].size;
            OE_READ_IN_OUT_PARAM(dc[_i_1].buf, dc[_i_1].size);
        }
    }
```